### PR TITLE
[WIP] Changesets

### PR DESCRIPTION
**Pull request by SlowMo24**
Add functionality for Changesets (addresses #52 )

### CODE REVIEW

File: oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/OSMChangeset.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import org.heigit.bigspatialdata.oshdb.util.OSHDBBoundingBox;
+import org.heigit.bigspatialdata.oshdb.util.OSHDBTimestamp;
+
+public class OSMChangeset {
+  private final long id;
+  private final Long user_id;
```
Comment (rtroilo) [inline oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/OSMChangeset.java]: should be the primitive type long and not Long Object type!

File: oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/OSMChangeset.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+  private final OSHDBTimestamp created_at;
+  private final OSHDBBoundingBox bbx;
+  private final OSHDBTimestamp closed_at;
+  private final Boolean open;
+  private final Integer num_changes;
```
Comment (rtroilo) [inline oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/OSMChangeset.java]: same as user_id, should be the primitive int type!

File: oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/OSMChangeset.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+import org.heigit.bigspatialdata.oshdb.util.OSHDBBoundingBox;
+import org.heigit.bigspatialdata.oshdb.util.OSHDBTimestamp;
+
+public class OSMChangeset {
+  private final long id;
+  private final Long user_id;
+  private final OSHDBTimestamp created_at;
+  private final OSHDBBoundingBox bbx;
+  private final OSHDBTimestamp closed_at;
+  private final Boolean open;
```
Comment (rtroilo) [inline oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/OSMChangeset.java]: also primitive type please

File: oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/OSMChangesetComment.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
+package org.heigit.bigspatialdata.oshdb.util.changesetenricher;
+
+import java.util.Objects;
+import org.heigit.bigspatialdata.oshdb.util.OSHDBTimestamp;
+
+public class OSMChangesetComment {
+  private final long comment_changeset_id;
+  private final Long comment_user_id;
```
Comment (rtroilo) [inline oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/OSMChangesetComment.java]: use primitive type for comment_user_id

File: oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/ChangesetEnricher.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.heigit.bigspatialdata.oshdb.util.OSHDBBoundingBox;
+import org.heigit.bigspatialdata.oshdb.util.OSHDBTimestamp;
+import org.slf4j.LoggerFactory;
+
+public class ChangesetEnricher {
+  private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(ChangesetEnricher.class);
```
Comment (rtroilo) [inline oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/ChangesetEnricher.java]: any reason not to import org.slf4j.Logger?

File: oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/ChangesetEnricher.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        + "WHERE id = ? ;")) {
+      prepareStatement.setLong(1, changesetId);
+      ResultSet resultSet = prepareStatement.executeQuery();
+
+      if (resultSet.next()) {
+        Long user_id = resultSet.getLong("user_id");
+
+        OSHDBTimestamp created_at;
+        try {
+          created_at = new OSHDBTimestamp(resultSet.getTimestamp("created_at"));
```
Comment (rtroilo) [inline oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/ChangesetEnricher.java]: don't use NullPointerException to check for null!
Please test explicit for it like 
```java
Timestamp ts = resultSet.getTimestamp("created_at");
if(ts != null) {
  created_at = new OSHDBTimestamp(ts);
}else {
 created_at = null;
}
```

File: oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/ChangesetEnricher.java [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        } catch (NullPointerException ex) {
+          created_at = null;
+        }
+
+        OSHDBBoundingBox bbx;
+        try {
+          bbx = new OSHDBBoundingBox(resultSet.getDouble("min_lon"), resultSet.getDouble("min_lat"), resultSet.getDouble("max_lon"), resultSet.getDouble("max_lat"));
```
Comment (rtroilo) [inline oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/ChangesetEnricher.java]: same as created_at, use explicit null checking!

File: oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/ChangesetEnricher.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+        } catch (NullPointerException ex) {
+          bbx = null;
+        }
+
+        OSHDBTimestamp closed_at;
+        try {
+          closed_at = new OSHDBTimestamp(resultSet.getTimestamp("closed_at"));
```
Comment (rtroilo) [inline oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/ChangesetEnricher.java]: see created_at comment

File: oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/ChangesetEnricher.java [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+   * This class works as a cache (like
+   * {@link org.heigit.bigspatialdata.oshdb.util.tagtranslator.TagTranslator}). Attention: The
+   * ChangesetDatabase is minutely updated but the cache isn't. So subsequent calls to this Class
+   * for the same ID will always return the same result. If you want to update the cache you will
+   * have to create a new instance.
+   *
+   * @param conn The connection to the ohsome-changeset postgres-database.
+   */
+  public ChangesetEnricher(Connection conn) {
+    this.changests = new ConcurrentHashMap<>(0);
```
Comment (rtroilo) [inline oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/ChangesetEnricher.java]: please explain why you use here ConcurrentHashMap here

File: oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/OSMChangeset.java (hunk 1 above)
Comment (SlowMo24) [reply oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/OSMChangeset.java]: For now we use this tool: https://github.com/ToeBee/ChangesetMD that clearly states:
> Note that all fields except for id and created_at can be null.

Therefore we have to use Objects, I guess

File: oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/OSMChangeset.java (hunk 3 above)
Comment (SlowMo24) [reply oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/OSMChangeset.java]: see above

File: oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/ChangesetEnricher.java (hunk 5 above)
Comment (SlowMo24) [reply oshdb-util/src/main/java/org/heigit/bigspatialdata/oshdb/util/changesetenricher/ChangesetEnricher.java]: thanks for spotting, was not handling my IDE correctly

**Comment by SlowMo24** [comment]
> you used different naming schemas snake case and camel case. I prefer came case! You should try to stick to one of them.

Yes, jenkins-codestyle also reported. -> fixed.

**Comment by SlowMo24** [comment]
TODO:
- [ ] write documentation
- [ ] think about easy access for OSHDB-Users

**Comment by tyrasd** [comment]
This has a new home at https://gitlab.gistools.geog.uni-heidelberg.de/giscience/big-data/ohsome/libs/changeset-metadata :house_with_garden:

**Pull request closed without merging**